### PR TITLE
Bugfix/auto normalizing logits binary

### LINKED ISSUE #1604: BinaryAccuracy() sometimes gives incorrect answers due to non-deterministic sigmoiding
## 🐛 Bug

`torchmetrics.classification.BinaryAccuracy` will apply a sigmoid to some inputs but not others leading to incorrect behavior.

## Details
The current behavior of BinaryAccuracy() is to apply a sigmoid transformation if the inputs are outside of [0, 1] before binarizing   

> If preds is a floating point tensor with values outside [0,1] range we consider the input to be logits and will auto apply sigmoid per element. 

i.e. 
y_hat = 1(sigmoid(z) >= threshold) if z outside [0, 1] 
y_hat = 1(z >= threshold) if z inside [0, 1] 

I assume z inside [0, 1] is checked for then entire batch (i.e. if one element of the batch is outside [0, 1] then we apply the sigmoid to everyone).

This will cause silent errors. In particular, if the user inputs logits then they expect the logits to always be sigmoided. However, it is totally possible for all of the logits to lie in [0, 1] for some batches in which case the input will **not** be sigmoided which will cause incorrect thresholding.

### To Reproduce

Here is a simple example. Support our network outputs logits.

```python
from torchmetrics.classification import BinaryAccuracy
from scipy.special import expit # expit = sigmoid
import numpy as np
import torch
```

This example should lead to a correct prediction
``` python
probability_thresh = 0.5 
logits = np.array([0.49]) # network output
target = np.array([1])

# logits of 0.49 give a probability of 0.62 indicating class 1, the correct prediction
expit(logits)
array([0.62010643])
int(expit(logits) >= probability_thresh) == target
True
```

`BinaryAccuracy()` however thinks it's an incorrect prediction~
``` python
# torchmetrics, however, thinks we have the inccorect prediction because it does NOT sigmoid the logits
ba = BinaryAccuracy(threshold=probability_thresh) 
ba.forward(preds=torch.tensor(logits), target=torch.tensor(target))
tensor(0.)
```

### Suggested Fix

I suggest adding an argument indicating whether or not the input predictions are sigmoided so the inputs are either always sigmoided or never sigmoided

### PR DESCRIPTION
**Pull request by JanFidor**
## What does this PR do?

Fixes #1604  for binary metrics

<details>
  <summary>Details </summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

The bug has been discussed a lot up to this point. I've decided to tackle it in parts, as merging the entire thing at one would be challenging. This PR fixes the behaviour in binary metrics. When we decide on the final nature of the fix and merge it, I will tackle the other ones.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

A LOT OF FUN, the day of the end of this vengeful bug draws nearer 

Credit to @SkafteNicki  who began the work on this bug and whose code this PR was heavily based on

### CODE REVIEW

**Comment by JanFidor** [comment]
Hi @SkafteNicki @justusschock  @Borda did you have a moment to take a lot the PR? I totally get it if you're busy, but if possible would very much appreciate it if you let me know when you could get around to a review! Have a great week! :sunny:

**Comment by JanFidor** [comment]
Hi again @SkafteNicki @justusschock @Borda any chance you might be able to do a review for this fix anytime soon? I'd love to make suitable fixes for the other metrics as soon as we merge this one.

**Comment by JanFidor** [comment]
I checked the maintainer list over at https://lightning.ai/docs/torchmetrics/stable/governance.html , so I'm moving my question to the currently listed maintainers. Hi @rittik9 @baskrahmer @matsumotosan , could you checkout this PR? I firmly believe that It would be a great addition to the library, as this bug is quite difficult to discover by oneself and as such could have silent impact on peoples workflows. Have amazing days! :sunny:
